### PR TITLE
SWATCH-598 Add worker for event messages to swatch tally

### PR DESCRIPTION
**Pull request by kartikshahc**
Jira issue: [SWATCH-598](https://issues.redhat.com/browse/SWATCH-598)

## Description

- This is an implementation of batch Kafka listener. swatch-tally listens on `platform.rhsm-subscriptions.service-instance-ingress` for event messages, and invokes EventController.saveAll() on them. If the json object is in bad format then we skip and log it as warning and process rest of the event messages in the batch. 
### Setup

1. Clean up the Events table.
2. Execute the below respective payload values mentioned in each step so that it forms a batch and then run the application `DEV_MODE=true ./gradlew clean :bootRun `
3. Use the Big data Tools of Intellij as producer to test the consumer topic `platform.rhsm-subscriptions.service-instance-ingress`. Payloads are as below make sure to check each Test step on which payload to use:
image

Payload1: `{"sla": "Premium", "role": "osd", "org_id": "1","timestamp": "2023-05-02T00:00:00Z", "event_type": "snapshot_redhat.com:openshift_dedicated:cluster_hour", "expiration": "2023-05-02T01:00:00Z", "instance_id": "e3a62bd1-fd00-405c-9401-f2288808588d", "display_name": "automation_osd_cluster_e3a62bd1-fd00-405c-9401-f2288808588d", "event_source": "prometheus", "measurements": [{"uom": "Instance-hours", "value": 1.0}], "service_type": "OpenShift Cluster"}`

Payload2: `{"sla": "Premium", "role": "osd", "org_id": "2","timestamp": "2023-05-02T00:00:00Z", "event_type": "snapshot_redhat.com:openshift_dedicated:cluster_hour", "expiration": "2023-05-02T01:00:00Z", "instance_id": "e3a62bd1-fd00-405c-9401-f2288808588d", "display_name": "automation_osd_cluster_e3a62bd1-fd00-405c-9401-f2288808588d", "event_source": "prometheus", "measurements": [{"uom": "Instance-hours", "value": 1.0}], "service_type": "OpenShift Cluster"}`

Payload3: `{"sla": "Premium", "role": "osd", "org_id": "3","timestamp": "2023-05-02T00:00:00Z", "event_type": "snapshot_redhat.com:openshift_dedicated:cluster_hour", "expiration": "2023-05-02T01:00:00Z", "instance_id": "e3a62bd1-fd00-405c-9401-f2288808588d", "display_name": "automation_osd_cluster_e3a62bd1-fd00-405c-9401-f2288808588d", "event_source": "prometheus", "measurements": [{"uom": "Instance-hours", "value": 1.0}], "service_type": "OpenShift Cluster"}`

Payload144: `{"sla": "Premium", "role": "osd", "org_id": "144","timestamp": "2023-05-02T00:00:00Z", "event_type": "snapshot_redhat.com:openshift_dedicated:cluster_hour", "expiration": "2023-05-02T01:00:00Z", "instance_id": "e3a62bd1-fd00-405c-9401-f2288808588d", "display_name": "automation_osd_cluster_e3a62bd1-fd00-405c-9401-f2288808588d", "event_source": "prometheus", "measurements": [{"uom": "Instance-hours", "value": 1.0}], "service_type": "OpenShift Cluster"}`

Payload145: `{"sla": "Premium1", "role": "osd", "org_id": "145","timestamp": "2023-05-02T00:00:00Z", "event_type": "snapshot_redhat.com:openshift_dedicated:cluster_hour", "expiration": "2023-05-02T01:00:00Z", "instance_id": "e3a62bd1-fd00-405c-9401-f2288808588d", "display_name": "automation_osd_cluster_e3a62bd1-fd00-405c-9401-f2288808588d", "event_source": "prometheus", "measurements": [{"uom": "Instance-hours", "value": 1.0}], "service_type": "OpenShift Cluster"}`

Payload146: `{"sla": "Premium", "role": "osd", "org_id": "146","timestamp": "2023-05-02T00:00:00Z", "event_type": "snapshot_redhat.com:openshift_dedicated:cluster_hour", "expiration": "2023-05-02T01:00:00Z", "instance_id": "e3a62bd1-fd00-405c-9401-f2288808588d", "display_name": "automation_osd_cluster_e3a62bd1-fd00-405c-9401-f2288808588d", "event_source": "prometheus", "measurements": [{"uom": "Instance-hours", "value": 1.0}], "service_type": "OpenShift Cluster"}`

### Steps

1. **Step 1: Add new events**
In the value string add payload 1, payload 2, payload 3 one by one and produce.

### Verification

1. New `three` entries created in table if the records don't already exists. Also in the log message you can see batch of three events accepted  `[INFO ] [org.candlepin.subscriptions.tally.billing.ServiceInstanceMessageConsumer] - Events received w/ event list size=3. Consuming events.`

```
fcac0338-15c7-4a74-aadd-a7f28fafab2e,,2023-05-02 00:00:00.000000 +00:00,"{""sla"": ""Premium"", ""role"": ""osd"", ""org_id"": ""2"", ""event_id"": ""fcac0338-15c7-4a74-aadd-a7f28fafab2e"", ""timestamp"": ""2023-05-02T00:00:00Z"", ""event_type"": ""snapshot_redhat.com:openshift_dedicated:cluster_hour"", ""expiration"": ""2023-05-02T01:00:00Z"", ""instance_id"": ""e3a62bd1-fd00-405c-9401-f2288808588d"", ""display_name"": ""automation_osd_cluster_e3a62bd1-fd00-405c-9401-f2288808588d"", ""event_source"": ""prometheus"", ""measurements"": [{""uom"": ""Instance-hours"", ""value"": 1.0}], ""service_type"": ""OpenShift Cluster""}",snapshot_redhat.com:openshift_dedicated:cluster_hour,prometheus,e3a62bd1-fd00-405c-9401-f2288808588d,2
552d14b4-7535-4d5f-9126-3935efa110e3,,2023-05-02 00:00:00.000000 +00:00,"{""sla"": ""Premium"", ""role"": ""osd"", ""org_id"": ""3"", ""event_id"": ""552d14b4-7535-4d5f-9126-3935efa110e3"", ""timestamp"": ""2023-05-02T00:00:00Z"", ""event_type"": ""snapshot_redhat.com:openshift_dedicated:cluster_hour"", ""expiration"": ""2023-05-02T01:00:00Z"", ""instance_id"": ""e3a62bd1-fd00-405c-9401-f2288808588d"", ""display_name"": ""automation_osd_cluster_e3a62bd1-fd00-405c-9401-f2288808588d"", ""event_source"": ""prometheus"", ""measurements"": [{""uom"": ""Instance-hours"", ""value"": 1.0}], ""service_type"": ""OpenShift Cluster""}",snapshot_redhat.com:openshift_dedicated:cluster_hour,prometheus,e3a62bd1-fd00-405c-9401-f2288808588d,3
12e52117-e7bf-4d97-9580-d3d0349cc2a5,,2023-05-02 00:00:00.000000 +00:00,"{""sla"": ""Premium"", ""role"": ""osd"", ""org_id"": ""1"", ""event_id"": ""12e52117-e7bf-4d97-9580-d3d0349cc2a5"", ""timestamp"": ""2023-05-02T00:00:00Z"", ""event_type"": ""snapshot_redhat.com:openshift_dedicated:cluster_hour"", ""expiration"": ""2023-05-02T01:00:00Z"", ""instance_id"": ""e3a62bd1-fd00-405c-9401-f2288808588d"", ""display_name"": ""automation_osd_cluster_e3a62bd1-fd00-405c-9401-f2288808588d"", ""event_source"": ""prometheus"", ""measurements"": [{""uom"": ""Instance-hours"", ""value"": 1.0}], ""service_type"": ""OpenShift Cluster""}",snapshot_redhat.com:openshift_dedicated:cluster_hour,prometheus,e3a62bd1-fd00-405c-9401-f2288808588d,1
```

### Steps

2. **Step 2: Update events**
Since we want the event_id for this step, in the value string copy the data column from the events table and then change a parameter value in it. I changed the event_source to `cost-management` and then reexecute it.
`{"sla": "Premium", "role": "osd", "org_id": "1", "event_id": "12e52117-e7bf-4d97-9580-d3d0349cc2a5", "timestamp": "2023-05-02T00:00:00Z", "event_type": "snapshot_redhat.com:openshift_dedicated:cluster_hour", "expiration": "2023-05-02T01:00:00Z", "instance_id": "e3a62bd1-fd00-405c-9401-f2288808588d", "display_name": "automation_osd_cluster_e3a62bd1-fd00-405c-9401-f2288808588d", "event_source": "cost-management", "measurements": [{"uom": "Instance-hours", "value": 1.0}], "service_type": "OpenShift Cluster"}`

### Verification

2. Updated `one` entry and since the application was already running as soon as Kafka listener received a batch it executed a single payload. `[INFO ] [org.candlepin.subscriptions.tally.billing.ServiceInstanceMessageConsumer] - Events received w/ event list size=1.`

`12e52117-e7bf-4d97-9580-d3d0349cc2a5,,2023-05-02 00:00:00.000000 +00:00,"{""sla"": ""Premium"", ""role"": ""osd"", ""org_id"": ""1"", ""event_id"": ""12e52117-e7bf-4d97-9580-d3d0349cc2a5"", ""timestamp"": ""2023-05-02T00:00:00Z"", ""event_type"": ""snapshot_redhat.com:openshift_dedicated:cluster_hour"", ""expiration"": ""2023-05-02T01:00:00Z"", ""instance_id"": ""e3a62bd1-fd00-405c-9401-f2288808588d"", ""display_name"": ""automation_osd_cluster_e3a62bd1-fd00-405c-9401-f2288808588d"", ""event_source"": ""cost-management"", ""measurements"": [{""uom"": ""Instance-hours"", ""value"": 1.0}], ""service_type"": ""OpenShift Cluster""}",snapshot_redhat.com:openshift_dedicated:cluster_hour,cost-management,e3a62bd1-fd00-405c-9401-f2288808588d,1`

### Steps

3. **Step 3: Update events like 21 new payloads and see it performs in batch**
Stop application. Update the same payload by changing org_ids and then reexecute the payloads. Once that is done restart the application.

Payload example: `{"sla": "Premium", "role": "osd", "org_id": "20", "event_id": "12e52117-e7bf-4d97-9580-d3d0349cc2a5", "timestamp": "2023-05-02T00:00:00Z", "event_type": "snapshot_redhat.com:openshift_dedicated:cluster_hour", "expiration": "2023-05-02T01:00:00Z", "instance_id": "e3a62bd1-fd00-405c-9401-f2288808588d", "display_name": "automation_osd_cluster_e3a62bd1-fd00-405c-9401-f2288808588d", "event_source": "cost-management", "measurements": [{"uom": "Instance-hours", "value": 1.0}], "service_type": "OpenShift Cluster"}
`
### Verification

3. In the log `Events received w/ event list size=15. Consuming events.`

### Steps

4. **Step 4: Add new events if its valid and skip the bad payloads**
Stop the service. In the value string add payload 144, payload 145, payload 146 one by one and produce. Notice that payload 145 is having bad data in sla as `Premium1`. Rerun the `DEV_MODE=true ./gradlew clean :bootRun `

### Verification

4. The batch will persist 144,146 payload by adding `two` entries and will log warning for 145 payload. 

Logs 
```
2023-07-11 18:09:30,723 [thread=swatch-instance-ingress-0-C-1] [INFO ] [org.candlepin.subscriptions.tally.billing.ServiceInstanceMessageConsumer] - Events received w/ event list size=3. Consuming events.
2023-07-11 18:09:30,924 [thread=swatch-instance-ingress-0-C-1] [WARN ] [org.candlepin.subscriptions.event.EventController] - Issue found com.fasterxml.jackson.databind.exc.ValueInstantiationException: Cannot construct instance of `org.candlepin.subscriptions.json.Event$Sla`, problem: Premium1
 at [Source: (String)"{"sla": "Premium1", "role": "osd", "org_id": "145","timestamp": "2023-05-02T00:00:00Z", "event_type": "snapshot_redhat.com:openshift_dedicated:cluster_hour", "expiration": "2023-05-02T01:00:00Z", "instance_id": "e3a62bd1-fd00-405c-9401-f2288808588d", "display_name": "automation_osd_cluster_e3a62bd1-fd00-405c-9401-f2288808588d", "event_source": "prometheus", "measurements": [{"uom": "Instance-hours", "value": 1.0}], "service_type": "OpenShift Cluster"}"; line: 1, column: 9] (through reference chain: org.candlepin.subscriptions.json.Event["sla"]) for the service instance json skipping to next
```

Tables with new entries look like below:

```
bd9fefa4-97ae-44f8-85f8-8206da1d37a2,,2023-05-02 00:00:00.000000 +00:00,"{""sla"": ""Premium"", ""role"": ""osd"", ""org_id"": ""144"", ""event_id"": ""bd9fefa4-97ae-44f8-85f8-8206da1d37a2"", ""timestamp"": ""2023-05-02T00:00:00Z"", ""event_type"": ""snapshot_redhat.com:openshift_dedicated:cluster_hour"", ""expiration"": ""2023-05-02T01:00:00Z"", ""instance_id"": ""e3a62bd1-fd00-405c-9401-f2288808588d"", ""display_name"": ""automation_osd_cluster_e3a62bd1-fd00-405c-9401-f2288808588d"", ""event_source"": ""prometheus"", ""measurements"": [{""uom"": ""Instance-hours"", ""value"": 1.0}], ""service_type"": ""OpenShift Cluster""}",snapshot_redhat.com:openshift_dedicated:cluster_hour,prometheus,e3a62bd1-fd00-405c-9401-f2288808588d,144
b9594613-47ae-4ee4-81e7-71f58bd8d8fa,,2023-05-02 00:00:00.000000 +00:00,"{""sla"": ""Premium"", ""role"": ""osd"", ""org_id"": ""146"", ""event_id"": ""b9594613-47ae-4ee4-81e7-71f58bd8d8fa"", ""timestamp"": ""2023-05-02T00:00:00Z"", ""event_type"": ""snapshot_redhat.com:openshift_dedicated:cluster_hour"", ""expiration"": ""2023-05-02T01:00:00Z"", ""instance_id"": ""e3a62bd1-fd00-405c-9401-f2288808588d"", ""display_name"": ""automation_osd_cluster_e3a62bd1-fd00-405c-9401-f2288808588d"", ""event_source"": ""prometheus"", ""measurements"": [{""uom"": ""Instance-hours"", ""value"": 1.0}], ""service_type"": ""OpenShift Cluster""}",snapshot_redhat.com:openshift_dedicated:cluster_hour,prometheus,e3a62bd1-fd00-405c-9401-f2288808588d,146
```

### CODE REVIEW

File: src/main/java/org/candlepin/subscriptions/metering/service/prometheus/config/PrometheusServiceConfiguration.java [hunk 1]
```diff
@@ -61,7 +61,7 @@ public PrometheusService prometheusService(ApiProvider provider) {
 
   @Bean
   EventController prometheusEventController(EventRecordRepository repo) {
-    return new EventController(repo);
+    return new EventController(repo, null);
```
Comment (mstead) [inline src/main/java/org/candlepin/subscriptions/metering/service/prometheus/config/PrometheusServiceConfiguration.java:64]: Passing null here will cause NPE in the event controller wherever this configuration is being used. It should be properly instantiated.

**Review by mstead — CHANGES_REQUESTED** [review]
Overall this looks good and works as described. One small thing I noticed.

File: src/main/java/org/candlepin/subscriptions/metering/service/prometheus/config/PrometheusServiceConfiguration.java (hunk 1 above)
Comment (kartikshahc) [reply src/main/java/org/candlepin/subscriptions/metering/service/prometheus/config/PrometheusServiceConfiguration.java]: Done

**Comment by ssidhaye** [comment]
After adding events via kafka-ui utility I am able to see the messages are being consumed in the tally-service pod logs

```
2023-07-12 13:55:16,249 [thread=swatch-instance-ingress-0-C-1] [INFO ] [org.candlepin.subscriptions.tally.billing.ServiceInstanceMessageConsumer] - Events received w/ event list size=1. Consuming events.
2023-07-12 13:56:09,970 [thread=swatch-instance-ingress-0-C-1] [INFO ] [org.candlepin.subscriptions.tally.billing.ServiceInstanceMessageConsumer] - Events received w/ event list size=1. Consuming events.
2023-07-12 13:56:27,645 [thread=swatch-instance-ingress-0-C-1] [INFO ] [org.candlepin.subscriptions.tally.billing.ServiceInstanceMessageConsumer] - Events received w/ event list size=1. Consuming events.
2023-07-12 13:56:39,049 [thread=swatch-instance-ingress-0-C-1] [INFO ] [org.candlepin.subscriptions.tally.billing.ServiceInstanceMessageConsumer] - Events received w/ event list size=1. Consuming events.
2023-07-12 13:56:39,053 [thread=swatch-instance-ingress-0-C-1] [WARN ] [org.candlepin.subscriptions.event.EventController] - Issue found com.fasterxml.jackson.databind.exc.ValueInstantiationException: Cannot construct instance of `org.candlepin.subscriptions.json.Event$Sla`, problem: Premium1
 at [Source: (String)"{"sla": "Premium1", "role": "osd", "org_id": "145","timestamp": "2023-05-02T00:00:00Z", "event_type": "snapshot_redhat.com:openshift_dedicated:cluster_hour", "expiration": "2023-05-02T01:00:00Z", "instance_id": "e3a62bd1-fd00-405c-9401-f2288808588d", "display_name": "automation_osd_cluster_e3a62bd1-fd00-405c-9401-f2288808588d", "event_source": "prometheus", "measurements": [{"uom": "Instance-hours", "value": 1.0}], "service_type": "OpenShift Cluster"}"; line: 1, column: 9] (through reference chain: org.candlepin.subscriptions.json.Event["sla"]) for the service instance json skipping to next
2023-07-12 13:57:02,181 [thread=swatch-instance-ingress-0-C-1] [INFO ] [org.candlepin.subscriptions.tally.billing.ServiceInstanceMessageConsumer] - Events received w/ event list size=1. Consuming events.
2023-07-12 14:01:43,451 [thread=swatch-instance-ingress-0-C-1] [INFO ] [org.candlepin.subscriptions.tally.billing.ServiceInstanceMessageConsumer] - Events received w/ event list size=1. Consuming events.
```



After updating an event

```
2023-07-12 14:04:30,151 [thread=swatch-instance-ingress-0-C-1] [INFO ] [org.candlepin.subscriptions.tally.billing.ServiceInstanceMessageConsumer] - Events received w/ event list size=1. Consuming events.
2023-07-12 14:05:04,502 [thread=swatch-instance-ingress-0-C-1] [INFO ] [org.candlepin.subscriptions.tally.billing.ServiceInstanceMessageConsumer] - Events received w/ event list size=1. Consuming events
```



Bad Payloads:

```
2023-07-12 13:56:39,053 [thread=swatch-instance-ingress-0-C-1] [WARN ] [org.candlepin.subscriptions.event.EventController] - Issue found com.fasterxml.jackson.databind.exc.ValueInstantiationException: Cannot construct instance of `org.candlepin.subscriptions.json.Event$Sla`, problem: Premium1
 at [Source: (String)"{"sla": "Premium1", "role": "osd", "org_id": "145","timestamp": "2023-05-02T00:00:00Z", "event_type": "snapshot_redhat.com:openshift_dedicated:cluster_hour", "expiration": "2023-05-02T01:00:00Z", "instance_id": "e3a62bd1-fd00-405c-9401-f2288808588d", "display_name": "automation_osd_cluster_e3a62bd1-fd00-405c-9401-f2288808588d", "event_source": "prometheus", "measurements": [{"uom": "Instance-hours", "value": 1.0}], "service_type": "OpenShift Cluster"}"; line: 1, column: 9] (through reference chain: org.candlepin.subscriptions.json.Event["sla"]) for the service instance json skipping to next
```



Update a bunch of events:


```
2023-07-12 15:40:33,951 [thread=main] [WARN ] [org.springframework.boot.autoconfigure.orm.jpa.JpaBaseConfiguration$JpaWebConfiguration] - spring.jpa.open-in-view is enabled by default. Therefore, database queries may be performed during view rendering. Explicitly configure spring.jpa.open-in-view to disable this warning
2023-07-12 15:40:36,851 [thread=Finalizer] [WARN ] [org.jboss.resteasy.client.jaxrs.i18n] - RESTEASY004687: Closing a class org.jboss.resteasy.client.jaxrs.engines.ApacheHttpClient43Engine instance for you. Please close clients yourself.
2023-07-12 15:40:44,896 [thread=main] [INFO ] [org.candlepin.subscriptions.BootApplication] - Started BootApplication in 44.674 seconds (JVM running for 46.58)
2023-07-12 15:40:48,524 [thread=swatch-instance-ingress-0-C-1] [INFO ] [org.candlepin.subscriptions.tally.billing.ServiceInstanceMessageConsumer] - Events received w/ event list size=11. Consuming events.
```

**Pull request merged**